frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx

```diff
@@ -0,0 +1,332 @@
+
+  const widgetId = element.id
+  const widgetFormId = element.formId
+
+  const uploadTheFile = useCallback(
+    (file: File) => {
+      uploadFiles({
+        files: [file],
+        uploadClient,
+        widgetMgr,
+        widgetInfo: { id: widgetId, formId: widgetFormId },
+      }).then(({ successfulUploads }) => {
+        const upload = successfulUploads[0]
+        if (upload && upload.fileUrl.deleteUrl) {
+          setDeleteFileUrl(upload.fileUrl.deleteUrl)
+        }
+      })
+    },
+    [uploadClient, widgetMgr, widgetId, widgetFormId]
+  )
+
+  useEffect(() => {
+    // this first part is to ensure we prompt for getting the user's media devices
+    navigator.mediaDevices
+      .getUserMedia({ audio: true })
+      .then(() => {
+        RecordPlugin.getAvailableAudioDevices().then(devices => {
+          setAvailableAudioDevices(devices)
+          if (devices.length > 0) {
+            setActiveAudioDeviceId(devices[0].deviceId)
+          }
+        })
+      })
+      .catch(_err => {
+        setHasNoMicPermissions(true)
+      })
+  }, [])
+
+  const initializeWaveSurfer = useCallback(() => {
+    if (waveSurferRef.current === null) return
+
+    const ws = WaveSurfer.create({
+      container: waveSurferRef.current,
+      waveColor: theme.colors.primary,
+      progressColor: theme.colors.bodyText,
+      height: WAVEFORM_HEIGHT - 2 * WAVEFORM_PADDING,
+      barWidth: BAR_WIDTH,
+      barGap: BAR_GAP,
+      barRadius: BAR_RADIUS,
+      cursorWidth: CURSOR_WIDTH,
+    })
+
+    ws.on("timeupdate", time => {
+      setProgressTime(formatTime(time * 1000)) // get from seconds to milliseconds
+    })
+
+    ws.on("pause", () => {
+      forceRerender()
+    })
+
+    const recordPlugin = ws.registerPlugin(
+      RecordPlugin.create({
+        scrollingWaveform: false,
+        renderRecordedAudio: true,
+      })
+    )
+
+    recordPlugin.on("record-end", blob => {
+      const url = URL.createObjectURL(blob)
+      setRecordingUrl(url)
+
+      const file = new File([blob], "audio.wav", { type: blob.type })
+      uploadTheFile(file)
+
+      ws.setOptions({
+        // This color is hardcoded because I've spent an hour trying to get it to work
+        // with the theme and I'm giving up for now. The complications arise due to the
+        // way wavesurfer is using these colors. Specifically, the progress color "tints" the wave color
+        // and the opacity of this wave color is not being respected, so to make things work for now hardcoding
+        // this color that looks "okay" on both themes.
+        waveColor: "#A5A5AA",
+      })
+    })
+
+    recordPlugin.on("record-progress", time => {
+      setRecordingTime(formatTime(time))
+    })
+
+    setWavesurfer(ws)
+    setRecordPlugin(recordPlugin)
+
+    return () => {
+      if (ws) ws.destroy()
+    }
+  }, [theme, uploadTheFile])
+
+  useEffect(() => {
+    initializeWaveSurfer()
```

Comment:
not sure whether this is a thing with waveSurfer, but is there an equivalent `deregisterPlugin` to the `registerPlugin` in [line 155](https://github.com/streamlit/streamlit/pull/9404/files#diff-a7eb8abbe0fa76f9716f9f399ee823661f1a6c761fa6065ef32204746da10ad4R155) and, if yes, would it make sense to call it in form of useEffect's returned callback when the component unmounts?

Reply:
hmm i didn't see anything like that in the docs. i think when destroy is called it all goes poof, but i can double check

Reply:
good intuition! looks like there is a way to deregister a plugin. https://github.com/katspaugh/wavesurfer.js/discussions/3238#discussioncomment-7206183

let me try that!